cairo_zero/tests/src/kakarot/precompiles/generate_modexp_data.py

```diff
@@ -0,0 +1,219 @@
+
+    add_test_case(
+        f"modexp_random_inputs_{base_len}_{exp_len}_{modulus_len}__{test_id}",
+        input_data,
+    )
+
+
+@given(
+    data=st.data(),
+    base_len=st.integers(min_value=1, max_value=48),  # At least 1 byte
+    exp_len=st.integers(min_value=1, max_value=48),  # At least 1 byte
+    modulus_len=st.integers(min_value=1, max_value=48),  # At least 1 byte
+)
+@settings(max_examples=50)
+def generate_test_modexp_384_bits_non_empty_outputs(
+    data, base_len, exp_len, modulus_len
+):
+    base = data.draw(st.binary(min_size=base_len, max_size=base_len))
+    exp = data.draw(st.binary(min_size=exp_len, max_size=exp_len))
+    # Ensure modulus is not zero by requiring at least one non-zero byte
+    modulus = data.draw(
+        st.binary(min_size=modulus_len, max_size=modulus_len).filter(
+            lambda x: any(b != 0 for b in x)
+        )
+    )
+
+    input_data = create_input_data(base_len, exp_len, modulus_len, base, exp, modulus)
+
+    add_test_case(
+        f"modexp_non_empty_outputs_{base_len}_{exp_len}_{modulus_len}", input_data
+    )
+
+
+def generate_cairo_tests() -> str:
+    """Generate Cairo test code from collected test cases."""
+    cairo_tests = []
+
+    # Remove duplicate test cases based on the name
+    filtered_test_cases = [
+        dict(t) for t in {case["name"]: case for case in test_cases}.values()
+    ]
+
+    for case in filtered_test_cases:
+        # Convert bytes to hex array format
+        input_bytes = [f"0x{b:02x}" for b in case["input"]]
+        test_name = f"test_{case['name']}"
+        test_input = f"let input = array![{', '.join(input_bytes)}];"
+
+        if not case["is_err"]:
+            output_bytes = [f"0x{b:02x}" for b in case["output_data"]]
+
+            test = f"""
+    #[test]
+    fn {test_name}() {{
+        #[cairofmt::skip]
+        {test_input}
+        #[cairofmt::skip]
+        let expected_result = array![{', '.join(output_bytes)}];
+        let expected_gas = {case['output_gas']};
+
+        let (gas, result) = ModExp::exec(input.span()).unwrap();
+        assert_eq!(result, expected_result.span());
+        assert_eq!(gas, expected_gas);
+    }}"""
+            cairo_tests.append(test)
+            continue
+
+        # In case the test errors, due to an input length too large
+        # just assert result.is_err()
+        test = f"""
+        #[test]
+        fn {test_name}() {{
+            #[cairofmt::skip]
+            {test_input}
+            let result = ModExp::exec(input.span());
+            assert!(result.is_err());
+        }}
+        """
+        cairo_tests.append(test)
+
+    return "\n".join(cairo_tests)
+
+
+if __name__ == "__main__":
+    # Run the Hypothesis test
+    generate_test_modexp_384_bits_random_inputs()
+    generate_test_modexp_384_bits_non_empty_outputs()
+
+    # Generate and save the Cairo tests
+    print(f"Generated {len(test_cases)} test cases")
+    with open("generated_modexp_tests.cairo", "w") as f:
```

Comment:
I must have missed something: those generated tests are not in this PR ?
In a previous version, they were added

Reply:
Oh got it, they were added to modexp.cairo directly